### DISCUSSION TITLE
Using typer.prompt with choices from Enum doesn't display choices

### DISCUSSION THREAD

**Discussion by StFS** [fastapi/typer#1332] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import typer
from enum import Enum


class ChoiceEnum(Enum):
    option_one = "opt1"
    option_two = "opt2"
    option_three = "opt3"


app = typer.Typer()


@app.command()
def subcommand(argument_option: ChoiceEnum = typer.Option(ChoiceEnum.option_two.value,
                                                          prompt="argument option",
                                                          show_choices=True)):
    prompt_option: ChoiceEnum = typer.prompt("prompt option",
                                             ChoiceEnum.option_three.value,
                                             show_choices=True)
    print(f"Argument choice: {argument_option}")
    print(f"Prompt choice: {prompt_option}")


if __name__ == "__main__":
    app()



# The result of running the above script results in the following output:
#
# > argument option (opt1, opt2, opt3) [opt2]: 
# > prompt option [opt3]: 
# > Argument choice: ChoiceEnum.option_two
# > Prompt choice: opt3
```

### Description

Using the `typer.prompt` function isn't displaying the choices of an enum the same way that an Option does.

Am I doing something wrong here? Is there some other way I should be doing the prompt in order for it to display the choices as is done with the argument choices?

Also, why are the return values different (the prompt gives me the string "opt3", while the argument returns the enum value itself).

### Operating System

macOS

### Typer Version

0.6.1

### Python Version

Python 3.10.8

**Comment by StFS** [comment]
Managed to figure out that I needed to add `type=ChoiceEnum` to the prompt call to get the return value to be of the same type as the argument. Still doesn't fix showing the choices.

**Comment by StFS** [comment]
Managed to find a workaround for this by not using an enum and instead use the `click.Choice` type:

```
    click_choice = click.Choice(['opt1a', 'opt2a', 'opt3a'])
    click_prompt_option: click.Choice = typer.prompt("click prompt option",
                                                     "opt1a",
                                                     show_choices=True,
                                                     type=click_choice)
```

Still think that the `typer` behavior needs to be looked at to make the behavior a bit more predictable. Would think that label should be updated to `bug`.

**Comment by MShekow** [comment]
I solved the issue by creating a new `Choice` class for `(Int)Enum`s:

```python
class IntEnumChoice(click.Choice):
    def __init__(self, enum_type: EnumType, case_sensitive: bool = True) -> None:
        choices = [f"{value.value}: {value.name}" for value in enum_type]
        super().__init__(choices, case_sensitive)
        self._enum_type = enum_type

    def convert(self, value: t.Any, param: t.Optional["Parameter"], ctx: t.Optional["Context"]) -> t.Any:
        try:
            return self._enum_type(int(value))
        except ValueError:
            choices_str = ", ".join(map(repr, self.choices))
            self.fail(
                ngettext(
                    "{value!r} is not {choice}.",
                    "{value!r} is not one of {choices}.",
                    len(self.choices),
                ).format(value=value, choice=choices_str, choices=choices_str),
                param,
                ctx,
            )
```
You can easily adapt it for _string_ `Enum`s (just remove the type conversion to `int` in the `return self._enum_type(int(value))` line).

**Comment by alanwilter** [comment]
I'm looking for a `typer` way to replace:

```python
parser.add_argument('move', choices=['rock', 'paper', 'scissors'])
```
It would be great to have a similar feature in the core of `typer`.

**Comment by mattmess1221** [comment]
You might be interested in using the [questionary](https://pypi.org/project/questionary/) library.

To integrate it with click/typer, you can use [click-prompt](https://pypi.org/project/click-prompt/). Typer support requires writing a mixin for `Command`

```py
from enum import StrEnum, auto
from typing import Annotated

import click
import click_prompt
import typer.core


class Command(typer.core.TyperCommand):
    def __call__(self, *args, **kwargs) -> None:
        for p in self.params:
            if isinstance(p, click.Option) and isinstance(p.type, click.Choice):
                p.__class__ = click_prompt.ChoiceOption
        super().__call__(*args, **kwargs)

# begin main logic

main = Typer()


class Framework(StrEnum):
    django = auto()
    flask = auto()
    fastapi = auto()
    pyramid = auto()
    sanic = auto()


# set command class to Command mixin
@main.command(cls=Command)
def run(*, framework: Annotated[Framework, typer.Option(prompt="Please choose a framework")] = Framework.fastapi):
    print(f"You chose to use the {framework} framework.")


if __name__ == "__main__":
    main()
```
![image](https://github.com/user-attachments/assets/20e1df07-9a3a-4f44-b2e3-a2e1ba9943c1)

**Comment by YuriiMotov** [comment, marked as answer]
`typer.prompt` is just re-export of `click.prompt`.
And, to show choices it requires `type` parameter passed and it should be instance of `Choice`.

https://github.com/pallets/click/blob/81a482fbfdd5a553cf4704f822ade04c3b102fbc/src/click/termui.py#L69

So, straightforward solution is to pass `type` parameter as follows:
```
    prompt_option: ChoiceEnum = typer.prompt(
        ...
        type=click.Choice([e.value for e in ChoiceEnum]),
    )
```
I will probably also want to specify `value_proc=ChoiceEnum` to convert the value from `str` to `ChoiceEnum`.

```py
import click
import typer
from enum import Enum


class ChoiceEnum(Enum):
    option_one = "opt1"
    option_two = "opt2"
    option_three = "opt3"

app = typer.Typer()

@app.command()
def subcommand():
    prompt_option: ChoiceEnum = typer.prompt(
        "prompt option",
        ChoiceEnum.option_three.value,
        type=click.Choice([e.value for e in ChoiceEnum]),
        value_proc=ChoiceEnum,
        show_choices=True,
    )
    print(f"Prompt choice: {prompt_option}")

if __name__ == "__main__":
    app()
```

Try it:

```
$ python main.py
prompt option (opt1, opt2, opt3) [opt3]: opt2
Prompt choice: ChoiceEnum.option_two
```
